Create specs's page database (#157)
PrestaShop/prestashop-specs: back-office/advanced-parameters/database/advanced_parameters-database-sql_manager.md

# SPECIFICATIONS - ADVANCED PARAMETERS &gt; DATABASE &gt; SQL MANAGER

The SQL manager is a complex feature, which should be reserved for technical people who know how to explore a database using the SQL language. In return for this complexity, it can be extremely powerful and proves immensely useful to those who master it.

## 1.7.8 Version 

> The page is going to be migrated into a module. The module will be downloadable for merchants who need it.

Reviewer: Native module ? Not pre-installed ?

Author: It is not defined for now but it seems to be not pre-installed

> The feature to hide the password or the secure key will be removed since we can't fully hide them since with an alias the keys or passwords can be displayed.

Reviewer: Are you talking of that ?

![capture d'écran_2829](https://user-images.githubusercontent.com/13449658/96855592-3ba33d80-145d-11eb-8428-d165fdcd5f2a.png)

Author: Yes, @PierreRambaud was saying that it can be easily bypassed.

Reviewer: All right, while reading I thougt you were talking of a feature like a "Show / hide password" button :smile: 
Maybe it will be more clearer if you add a screenshot ? Or say directly that they will be displayed in clear?

## 1.7.7.0 Version

This tool enables you to perform SQL queries directly on the PrestaShop database, and save them for use at any later time. 

For instance, using this tool and your knowledge of SQL, you could create a reusable query giving you an updated list of the clients who are subscribed to your newsletter or get a list of products in HTML or CSV format.

> For security reasons, some types of queries are not allowed: UPDATE, DELETE, CREATE TABLE, DROP, etc. In short, you can only read data (SELECT query).

Reviewer 2: Unauthorized keywords:
        'unauthorized' => [
            'DELETE', 'ALTER', 'INSERT', 'REPLACE', 'CREATE', 'TRUNCATE', 'OPTIMIZE', 'GRANT', 'REVOKE', 'SHOW', 'HANDLER',
            'LOAD', 'ROLLBACK', 'SAVEPOINT', 'UNLOCK', 'INSTALL', 'UNINSTALL', 'ANALZYE', 'BACKUP', 'CHECK', 'CHECKSUM', 'REPAIR', 'RESTORE', 'CACHE',
            'DESCRIBE', 'EXPLAIN', 'USE', 'HELP', 'SET', 'DUPLICATE', 'VALUES',  'INTO', 'RENAME', 'CALL', 'PROCEDURE',  'FUNCTION', 'DATABASE', 'SERVER',
            'LOGFILE', 'DEFINER', 'RETURNS', 'EVENT', 'TABLESPACE', 'VIEW', 'TRIGGER', 'DATA', 'DO', 'PASSWORD', 'USER', 'PLUGIN', 'FLUSH', 'KILL',
            'RESET', 'START', 'STOP', 'PURGE', 'EXECUTE', 'PREPARE', 'DEALLOCATE', 'LOCK', 'USING', 'DROP', 'FOR', 'UPDATE', 'BEGIN', 'BY', 'ALL', 'SHARE',
            'MODE', 'TO', 'KEY', 'DISTINCTROW', 'DISTINCT',  'HIGH_PRIORITY', 'LOW_PRIORITY', 'DELAYED', 'IGNORE', 'FORCE', 'STRAIGHT_JOIN',
            'SQL_SMALL_RESULT', 'SQL_BIG_RESULT', 'QUICK', 'SQL_BUFFER_RESULT', 'SQL_CACHE', 'SQL_NO_CACHE', 'SQL_CALC_FOUND_ROWS', 'WITH',
        ],

At first, the SQL Manager page displays 2 messages:
- Alert info:

  "How do I create a new SQL query?
    - Click "Add New".
    - Fill in the fields and click "Save".
    - You can then view the query results by clicking on the Edit action in the dropdown menu
    - You can also export the query results as a CSV file by clicking on the Export button
"

- Alert warning:

"When saving the query, only the "SELECT" SQL statement is allowed."

### SQL Manager List

> Below the messages, the page displays the SQL Manager list with the request's ID, SQL query name, and action columns.

Reviewer: ```suggestion
Below the messages, the page displays the SQL Manager list with the request's ID, SQL query name, SQL query and actions columns.
```
```suggestion
Below the messages, the page displays the SQL Manager list with the request's ID, SQL query name, and action columns.
```

The request's actions allow us to download the request's result into a CSV file, view the request's result on the page, delete the request, and edit it.
To summarize the actions cab be:

> - Export. Performs the query, and have it downloaded by you in CSV format.

Reviewer: What do you mean by "performs the query" ? 
The query is just downloaded in CSV format

Author: Execute the query to retrieve the data and then download it. It is not the query that is exported but the results of the query. I have doubts now, I'm going to check

Reviewer: Ah, yes my bad you are right it's the results of the query !
So the query is well executed to get the results :)

# SPECIFICATIONS - ADVANCED PARAMETERS &gt; DATABASE &gt; SQL MANAGER

The SQL manager is a complex feature, which should be reserved for technical people who know how to explore a database using the SQL language. In return for this complexity, it can be extremely powerful and proves immensely useful to those who master it.

## 1.7.8 Version 

The page is going to be migrated into a module. The type of module is still to be defined, the most likely solution seems to be an not installed module.

> The module will be downloadable for merchants who need it and have the required skills.

Reviewer: ```suggestion
The module will be downloadable for users who need it and have the required skills.
```

For security reasons, some types of queries are not allowed: UPDATE, DELETE, CREATE TABLE, DROP, etc. In short, you can only read data (SELECT query).

At first, the SQL Manager page displays 2 messages:
- Alert info:

  "How do I create a new SQL query?
    - Click "Add New".
    - Fill in the fields and click "Save".
    - You can then view the query results by clicking on the Edit action in the dropdown menu
    - You can also export the query results as a CSV file by clicking on the Export button
"

- Alert warning:

"When saving the query, only the "SELECT" SQL statement is allowed."

### SQL Manager List

Below the messages, the page displays the SQL Manager list with the request's ID, SQL query name, SQL query and actions columns.
The request's actions allow us to download the request's result into a CSV file, view the request's result on the page, delete the request, and edit it.
To summarize the actions cab be:
- Export. Performs the query, and have it downloaded by you in CSV format.

> - View. Performs the query, and has it displayed in an HTML table, right within the PrestaShop interface.

Reviewer: ```suggestion
- View. Performs the query, and get the results displayed in an HTML table, right within the PrestaShop interface.
```

- Edit. You can edit a query as often as necessary, to refine it and get better results.
- Delete. Once a query is not used anymore (or simply because it does not work), you can delete it by clicking this button and confirming your choice.
When the user views the SQL request with no result, a warning is displayed: "This SQL query has no result."

### Settings

Select your default file encoding. You can configure the character encoding of the downloaded CSV file. The default, UTF-8, is recommended, but you can select ISO-8859-1 if need be.

### Add new SQL query

Creating a new query is divided into 3 sections: 

- The SQL form with 2 following inputs : 

    - SQL query name. Make the name as long and descriptive as necessary.
    - SQL query. The SQL query itself. You are free to perform JOINs or other intricate selections.
 The cancel button redirects the user to the SQL Manager list without saving any information.
 

>  The save button saves the SQL name and request and redirects the merchant to the list with the new request saved. If the merchant saves a request that has not the authorized querie type (defined in the list bellow), a error is displayed : "Your request is not authorised, please try a SELECT type request", the query is not saved and the user stays on the page with the displayed error.

Reviewer: ```suggestion
 The save button saves the SQL query name, the SQL query and redirects the user to the list with the new SQL query saved. If the user saves a request that has not the authorized query type (defined in the list bellow), an error is displayed : "Your request is not authorised, please try a SELECT type request", the query is not saved and the user stays on the page with the error displayed.
```

### SQL Manager List

Below the messages, the page displays the SQL Manager list with the request's ID, SQL query name, SQL query, and the actions columns.
The request's actions allow us to download the request's result into a CSV file, view the request's result on the page, delete the request, and edit it.
To summarize the actions cab be:
- Export. Executes the query, and downloaded the result in CSV format.
- View. Executes the query, and get the results displayed in an HTML table, right within the PrestaShop interface.

When the user views the SQL request with no result, a warning is displayed: "This SQL query has no result."
- Edit. The user can edit a query as often as necessary, to refine it and get better results.
- Delete. Once a query is not used anymore (or simply because it does not work), you can delete it by clicking this button and confirming your choice.

### Settings

Select your default file encoding. You can configure the character encoding of the downloaded CSV file. The default, UTF-8, is recommended, but you can select ISO-8859-1 if need be.

### Add new SQL query

Creating a new query is divided into 3 sections: 

> - The SQL form with 2 following inputs : 

Reviewer 3: ```suggestion
- The SQL form has 2 following input fields:
```

>     - SQL query name. Make the name as long and descriptive as necessary.

Reviewer 3: ```suggestion
    - SQL query name. Users can type whatever name they need, there is no limitation in the number or the characters allowed.
```

>     - SQL query. The SQL query itself. You are free to perform JOINs or other intricate selections.

Reviewer 3: What do we expect here?

The cancel button redirects the user to the SQL Manager list without saving any information.
 

> The save button saves the SQL query name, the SQL query, and redirects the user to the list adding the new SQL query saved with the success alert: "Successful creation.".

Reviewer 3: ```suggestion
The save button saves the SQL query name, the SQL query, and redirects the user to the list adding the new SQL query saved with the success notification: "Successful creation.".
```

> The save button saves the SQL query name, the SQL query, and redirects the user to the list adding the new SQL query saved with the success alert: "Successful creation.".

Reviewer 4: ```suggestion
The save button saves the SQL query name, the SQL query, and redirects the user to the list adding the new SQL query saved with the success alert: "SQL query was successfully saved."
```

> If the user saves a request that has not the authorized query type (defined in the list below), an error is displayed: "Your request is not authorized, please try a SELECT type request", the query is not saved and the user stays on the page with the error displayed.

Reviewer 4: ```suggestion
If the user saves a request that has not the authorized query type (defined in the list below), an error is displayed: "Error: only SELECT type queries are authorized.", the query is not saved and the user stays on the page with the error displayed.
```

The list of unauthorized request types:
'DELETE', 'ALTER', 'INSERT', 'REPLACE', 'CREATE', 'TRUNCATE', 'OPTIMIZE', 'GRANT', 'REVOKE', 'SHOW', 'HANDLER',
'LOAD', 'ROLLBACK', 'SAVEPOINT', 'UNLOCK', 'INSTALL', 'UNINSTALL', 'ANALZYE', 'BACKUP', 'CHECK', 'CHECKSUM', 'REPAIR', 'RESTORE', 'CACHE',
'DESCRIBE', 'EXPLAIN', 'USE', 'HELP', 'SET', 'DUPLICATE', 'VALUES', 'INTO', 'RENAME', 'CALL', 'PROCEDURE', 'FUNCTION', 'DATABASE', 'SERVER',
'LOGFILE', 'DEFINER', 'RETURNS', 'EVENT', 'TABLESPACE', 'VIEW', 'TRIGGER', 'DATA', 'DO', 'PASSWORD', 'USER', 'PLUGIN', 'FLUSH', 'KILL',
'RESET', 'START', 'STOP', 'PURGE', 'EXECUTE', 'PREPARE', 'DEALLOCATE', 'LOCK', 'USING', 'DROP', 'FOR', 'UPDATE', 'BEGIN', 'BY', 'ALL', 'SHARE',
'MODE', 'TO', 'KEY', 'DISTINCTROW', 'DISTINCT', 'HIGH_PRIORITY', 'LOW_PRIORITY', 'DELAYED', 'IGNORE', 'FORCE', 'STRAIGHT_JOIN',
'SQL_SMALL_RESULT', 'SQL_BIG_RESULT', 'QUICK', 'SQL_BUFFER_RESULT', 'SQL_CACHE', 'SQL_NO_CACHE', 'SQL_CALC_FOUND_ROWS', 'WITH'.
Each keywords are checked in the SQL query before saving it.

> On the bottom left, the "List of MySQL Tables" gives the user a list of all the available database tables. Select a table to make PrestaShop display its attributes and types on the list of attributes for this MySQL table, then click "Add attribute to SQL query" or "Add table name to SQL query" to send its attribute name or table name into the SQL query field.

Reviewer 3: ```suggestion
On the bottom left, the "List of MySQL Tables" gives the user a list of all the available database tables. He/she should select a table to make PrestaShop display all its attributes and types in the list on the right, then click "Add attribute to SQL query" or "Add table name to SQL query" to send its attribute name/table name into the SQL query field.
```

### SQL Manager List

Below the messages, the page displays the SQL Manager list with the request's ID, SQL query name, SQL query and actions columns.
The request's actions allow us to download the request's result into a CSV file, view the request's result on the page, delete the request, and edit it.
To summarize the actions cab be:

> - Export. Performs the query, and have it downloaded by you in CSV format.

Reviewer: ```suggestion
- Export. Performs the query, and have the query's results downloaded in CSV format.
```
